storage/replica.go

```diff
@@ -1558,3 +1557,23 @@ func (r *Replica) maybeAddToSplitQueue() {
 		r.store.splitQueue.MaybeAdd(r, r.store.Clock().Now())
 	}
 }
+
+// Quiesce drains the replica and prepares it for removal. First, proposal
+// of new commands is disabled. Secondly, all pending commands are aborted.
+// In both cases, a RangeNotFoundError results.
+func (r *Replica) Quiesce() {
+	r.Lock()
+	r.proposeRaftCommandFn = func(_ cmdIDKey, _ roachpb.RaftCommand) <-chan error {
+		ch := make(chan error, 1)
+		ch <- multiraft.ErrGroupDeleted
+		return ch
+	}
+	for id, cmd := range r.pendingCmds {
+		cmd.done <- roachpb.ResponseWithError{Err: multiraft.ErrGroupDeleted}
+		// Deleting while iterating is ok in Go. We don't just want to `nil`
+		// this since there could be commands being queued up for execution on
```

Comment:
But don't you want all commands which might be added to `pendingCmds` to fail? It looks somewhat fragile that we fail all currently-pending commands while allowing others to be written (with the expectation that they will fail when they reach the patched `proposeRaftCmdFn`. I think it would be better to add an explicit `quiesced` bool field to the `Replica` and use it in `proposeRaftCommand` (and then we should be able to simply nil out the `pendingCmds` map)